### PR TITLE
satellite/metainfo: Improve metainfo ListSegments

### PR DESCRIPTION
**Pull request by mniewrzal**
What: Improve metainfo `ListSegments` method with recent change about not encrypting a number of segments in metadata. With a known number of segments, we can just calculate result without many pointerDB calls. The implementation supports old and new way.

Why: We don't want to query pointerDB for each segment separately but only one if we have number of segments available (not encrypted).

Please describe the tests:
 - Test 1: automatic tests added 
 
Please describe the performance impact: newly created segments should be listed faster

## Code Review Checklist (to be filled out by reviewer)
 - [x] Does the PR describe what changes are being made?
 - [x] Does the PR describe why the changes are being made?
 - [x] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [x] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [x] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [x] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [x] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [x] Does any documentation need updating?
 - [x] Do the database access patterns make sense?

### CODE REVIEW

File: satellite/metainfo/metainfo.go [hunk 1]
```diff
@@ -1630,41 +1630,93 @@ func (endpoint *Endpoint) ListSegments(ctx context.Context, req *pb.SegmentListR
...
-	// TODO think about better implementation
-	for {
-		path, err := CreatePath(ctx, keyInfo.ProjectID, index, streamID.Bucket, streamID.EncryptedPath)
-		if err != nil {
-			return nil, status.Error(codes.Internal, err.Error())
+	more := false
+
+	if streamMeta.NumberOfSegments > 0 {
+		// use unencrypted number of segments
+		// TODO cleanup int32 vs int64
```
[the remark this answers is not in the document]
Comment (mniewrzal) [reply satellite/metainfo/metainfo.go]: Unfortunately, I'm not sure if this is scope for this ticket. We have different types for fields from the same group in different protobufs. In this case we have `streamMeta.NumberOfSegments` `int64` from `streams` proto and segment `Index` `int32` from `metainfo` proto. I will create ticket for that.

File: satellite/metainfo/metainfo.go [hunk 2]
```diff
@@ -1630,41 +1630,93 @@ func (endpoint *Endpoint) ListSegments(ctx context.Context, req *pb.SegmentListR
...
+				})
+			}
 		}
-		_, err = endpoint.metainfo.Get(ctx, path)
-		if err != nil {
-			if storage.ErrKeyNotFound.Has(err) {
-				if index == lastSegment {
+	} else {
+		index := int64(req.CursorPosition.Index)
+		// TODO think about better implementation
```
[the remark this answers is not in the document]
Comment (mniewrzal) [reply satellite/metainfo/metainfo.go]: after code reorganization now it looks better and I think its good enough to remove this `TODO`. Let me know what do you think.

File: satellite/metainfo/metainfo.go [hunk 3]
```diff
@@ -1630,41 +1630,93 @@ func (endpoint *Endpoint) ListSegments(ctx context.Context, req *pb.SegmentListR
...
+
 	segmentItems := make([]*pb.SegmentListItem, 0)
-	// TODO think about better implementation
-	for {
-		path, err := CreatePath(ctx, keyInfo.ProjectID, index, streamID.Bucket, streamID.EncryptedPath)
-		if err != nil {
-			return nil, status.Error(codes.Internal, err.Error())
+	more := false
+
+	if streamMeta.NumberOfSegments > 0 {
```
[the remark this answers is not in the document]
Comment (mniewrzal) [reply satellite/metainfo/metainfo.go]: both `if/else` blocks are now moved to separate functions

File: satellite/metainfo/metainfo.go [hunk 4]
```diff
@@ -1630,41 +1630,93 @@ func (endpoint *Endpoint) ListSegments(ctx context.Context, req *pb.SegmentListR
...
+		// use unencrypted number of segments
+		// TODO cleanup int32 vs int64
+		numberOfSegments := int32(streamMeta.NumberOfSegments) - req.CursorPosition.Index
+		if numberOfSegments > 0 {
+			if numberOfSegments > limit {
+				more = true
+				numberOfSegments = limit
+			} else {
+				// remove last segment from loop
+				numberOfSegments--
```
[the remark this answers is not in the document]
Comment (mniewrzal) [reply satellite/metainfo/metainfo.go]: done

File: satellite/metainfo/metainfo.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (isaachess) [reply satellite/metainfo/metainfo.go]: :+1:

**Pull request merged**